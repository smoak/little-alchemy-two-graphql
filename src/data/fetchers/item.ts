import low from 'lowdb';
import FileSync from 'lowdb/adapters/FileSync';

const adapter = new FileSync<DatabaseSchema>('db.json');
const db = low(adapter);

interface DatabaseSchema {
  readonly items: DatabaseItems;
}

export interface DatabaseItem {
  readonly myths: boolean;
  readonly makes: Record<string, string>;
  readonly combinations: string[][];
}

type DatabaseItems = Record<string, DatabaseItem>;

type GetItems = () => DatabaseItems;
export const getItems: GetItems = () => db.get('items').value();

type GetItemByName = (name: string) => DatabaseItem | undefined;
export const getItemByName: GetItemByName = name => db.get('items').value()[name];

type GetItemNames = () => string[];
export const getItemNames: GetItemNames = () => db.get('items').keys().value();
