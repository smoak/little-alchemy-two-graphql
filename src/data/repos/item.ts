import Fuse from 'fuse.js';

import { DatabaseItem, getItemByName, getItemNames, getItems } from '../fetchers/item';

export interface Item {
  readonly name: string;
  readonly myths: boolean;
  readonly creates: ItemCombination[];
  readonly combinations: ItemCombination[];
  readonly imageUrl: string;
}

export interface ItemCombination {
  readonly source: string;
  readonly target: string;
}

const fromCombinations = (combinations: DatabaseItem['combinations']): ItemCombination[] =>
  combinations.map(combo => ({ source: combo[0], target: combo[1] }));

const fromItemWithName = (name: string, item: DatabaseItem): Item => ({
  name,
  imageUrl: item.image_url,
  myths: item.myths,
  creates: item.makes,
  combinations: fromCombinations(item.combinations),
});

type FindById = (id: string) => Item;
export const findById: FindById = id => {
  const item = getItemByName(id);

  if (!item) {
    throw new Error(`Could not find item with id: ${id}`);
  }

  return fromItemWithName(id, item);
};

type FindAll = () => Item[];
export const findAll: FindAll = () => {
  const items = getItems();

  return Object.keys(items).map<Item>(itemName => {
    const item = items[itemName];
    return fromItemWithName(itemName, item);
  });
};

type Search = (query: string) => Item[];
export const search: Search = query => {
  const items = getItemNames();
  const options: Fuse.IFuseOptions<string> = {
    minMatchCharLength: 2,
    threshold: 1,
  };
  const fuse = new Fuse(items, options);
  const results = fuse.search(query);
  return results.map(result => findById(result.item));
};
