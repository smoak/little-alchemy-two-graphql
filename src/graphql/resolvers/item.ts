import { GraphQLFieldResolver } from 'graphql';

import { ItemCombination, findAll, findById } from '../../data/repos/item';
import { Query } from '../types';

export const ItemsResolver: GraphQLFieldResolver<Query, unknown, unknown> = (_source, _args, _context, _info) =>
  findAll();

interface ItemWithNameResolverArgs {
  readonly name: string;
}
export const ItemWithNameResolver: GraphQLFieldResolver<Query, unknown, ItemWithNameResolverArgs> = (_source, args) =>
  findById(args.name);

export const SourceItemCombinationResolver: GraphQLFieldResolver<ItemCombination, unknown, unknown> = (
  parent,
  _args,
  _context,
  _info
) => findById(parent.source);

export const TargetItemCombinationResolver: GraphQLFieldResolver<ItemCombination, unknown, unknown> = (
  parent,
  _args,
  _context,
  _info
) => findById(parent.target);
