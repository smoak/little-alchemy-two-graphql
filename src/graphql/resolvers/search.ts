import { Connection, ConnectionArguments, connectionFromArray } from 'graphql-relay';

import { Item, search } from '../../data/repos/item';
import { NoSearchResults, SearchArgs } from '../types';

export interface SearchResultsResolverItemsResults {
  readonly items: Item[];
}
export type SearchResultsResolverResults = NoSearchResults | SearchResultsResolverItemsResults;

type SearchResultsResolver = (parent: unknown, args: SearchArgs) => SearchResultsResolverResults;
export const searchResultsResolver: SearchResultsResolver = (_, { query }) => {
  const searchResults = search(query);

  if (searchResults.length === 0) {
    return {
      message: `No results matched ${query}`,
    };
  }

  return { items: searchResults };
};

type ItemSearchResultsResolver = (
  parent: SearchResultsResolverItemsResults,
  args: ConnectionArguments
) => Connection<Item>;
export const itemSearchResultsResolver: ItemSearchResultsResolver = (parent, args) =>
  connectionFromArray(parent.items, args);
