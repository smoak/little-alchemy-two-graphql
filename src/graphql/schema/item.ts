import {
  GraphQLBoolean,
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
  Thunk,
} from 'graphql';
import { globalIdField } from 'graphql-relay';

import { Item, ItemCombination } from '../../data/repos/item';
import { SourceItemCombinationResolver, TargetItemCombinationResolver } from '../resolvers';

import { nodeInterface } from './node';

interface ItemFields extends GraphQLFieldConfigMap<Item, unknown, unknown> {
  readonly name: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly myths: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly id: GraphQLFieldConfig<unknown, unknown, unknown>;
  readonly creates: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly combinations: GraphQLFieldConfig<Item, unknown, unknown>;
}

interface ItemCombinationFields extends GraphQLFieldConfigMap<ItemCombination, unknown, unknown> {
  readonly source: GraphQLFieldConfig<ItemCombination, unknown, unknown>;
  readonly target: GraphQLFieldConfig<ItemCombination, unknown, unknown>;
}

type Fields = () => ItemFields;
const fields: Fields = () => ({
  name: { type: new GraphQLNonNull(GraphQLString) },
  myths: { type: new GraphQLNonNull(GraphQLBoolean) },
  id: globalIdField('Item', obj => obj.name),
  creates: {
    description: 'The items that this item creates',
    type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(ItemCombinationType))),
  },
  combinations: {
    description: 'The items that create this item',
    type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(ItemCombinationType))),
  },
});

const itemCombinationFields: Thunk<ItemCombinationFields> = () => ({
  source: { type: new GraphQLNonNull(ItemType), resolve: SourceItemCombinationResolver },
  target: { type: new GraphQLNonNull(ItemType), resolve: TargetItemCombinationResolver },
});

export const ItemType = new GraphQLObjectType<Item, unknown, unknown>({
  name: 'Item',
  description: 'An item',
  interfaces: [nodeInterface],
  fields,
  isTypeOf: source =>
    Object.keys(source).includes('name') &&
    Object.keys(source).includes('myths') &&
    Object.keys(source).includes('creates'),
});

export const ItemCombinationType = new GraphQLObjectType<ItemCombination, unknown, unknown>({
  name: 'ItemCombination',
  description: 'A combination of two items',
  fields: itemCombinationFields,
});
