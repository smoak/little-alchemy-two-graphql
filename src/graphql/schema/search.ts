import { GraphQLFieldConfig, GraphQLNonNull, GraphQLObjectType, GraphQLString, GraphQLUnionType } from 'graphql';
import { connectionArgs } from 'graphql-relay';

import { itemSearchResultsResolver, searchResultsResolver } from '../resolvers/search';
import { SearchArgs } from '../types';

import { ItemConnection } from './item';

export const ItemSearchResultsType = new GraphQLObjectType({
  name: 'ItemSearchResults',
  description: 'Item search results',
  fields: {
    items: {
      description: 'A paginated list of items',
      type: new GraphQLNonNull(ItemConnection),
      args: connectionArgs,
      resolve: itemSearchResultsResolver,
    },
  },
  isTypeOf: source => Object.keys(source).includes('items'),
});

const NoSearchResultsType = new GraphQLObjectType({
  name: 'NoSearchResults',
  fields: {
    message: {
      description: 'A message indicating why the results could not be found',
      type: new GraphQLNonNull(GraphQLString),
    },
  },
  isTypeOf: source => Object.keys(source).includes('message'),
});

const SearchResultsType = new GraphQLUnionType({
  name: 'SearchResults',
  types: [NoSearchResultsType, ItemSearchResultsType],
});

export const searchField: GraphQLFieldConfig<unknown, unknown, SearchArgs> = {
  description: 'Search based on input',
  type: new GraphQLNonNull(SearchResultsType),
  args: {
    query: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The query to execute',
    },
  },
  resolve: searchResultsResolver,
};
