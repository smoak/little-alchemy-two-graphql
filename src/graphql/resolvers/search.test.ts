jest.mock('../../data/repos/item');

import { ConnectionArguments } from 'graphql-relay';

import { Item, search } from '../../data/repos/item';

import { SearchResultsResolverItemsResults, itemSearchResultsResolver, searchResultsResolver } from './search';

describe('search', () => {
  const items: Item[] = [{ name: 'test', myths: false, creates: [], combinations: [] }];

  describe('.searchResultsResolver', () => {
    let result: ReturnType<typeof searchResultsResolver>;

    describe('with no results', () => {
      beforeEach(() => {
        (search as jest.Mock).mockReturnValue([]);

        result = searchResultsResolver({}, { query: 'test' });
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });

    describe('with results', () => {
      beforeEach(() => {
        (search as jest.Mock).mockReturnValue(items);

        result = searchResultsResolver({}, { query: 'test' });
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('.itemSearchResultsResolver', () => {
    let result: ReturnType<typeof itemSearchResultsResolver>;

    beforeEach(() => {
      const parent: SearchResultsResolverItemsResults = {
        items,
      };
      const args: ConnectionArguments = {};
      result = itemSearchResultsResolver(parent, args);
    });

    it('returns the expected result', () => {
      expect(result).toMatchSnapshot();
    });
  });
});
