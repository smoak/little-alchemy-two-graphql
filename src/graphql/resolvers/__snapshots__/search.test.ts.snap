// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search .itemSearchResultsResolver returns the expected result 1`] = `
{
  "edges": [
    {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": {
        "combinations": [],
        "creates": [],
        "imageUrl": "https://picsum.photos/48",
        "myths": false,
        "name": "test",
      },
    },
  ],
  "pageInfo": {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjA=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`search .searchResultsResolver with no results returns the expected result 1`] = `
{
  "message": "No results matched test",
}
`;

exports[`search .searchResultsResolver with results returns the expected result 1`] = `
{
  "items": [
    {
      "combinations": [],
      "creates": [],
      "imageUrl": "https://picsum.photos/48",
      "myths": false,
      "name": "test",
    },
  ],
}
`;
